v21.1/cluster-api.md

```diff
@@ -0,0 +1,67 @@
+---
+title: Cluster API
+summary: Programmatically access and monitor cluster and node status information with a RESTful API.
+toc: true
+---
+
+The CockroachDB Cluster API is a REST API that provides information about a cluster and its nodes. The API offers programmatic access to much of the information available in the [DB Console](ui-overview.html) user interface, enabling you to monitor and troubleshoot your cluster using your choice of tooling.
+
+The Cluster API is hosted by all nodes of your cluster and provides information about all nodes. The API is available on the same port that is listening for HTTP connections to the DB Console.
+
+## Resources
+
+The following endpoints are available as URLs under the `/api/v2` base path (for example, `https://localhost:8080/api/v2/health/`). Additional endpoints are planned for future versions of CockroachDB.
+
+Each listed endpoint links to its full [API reference documentation](../api/cluster/v2.html).
+
+Endpoint | Name | Description
+--- | --- | ---
+[`/health`](../api/cluster/v2.html#operation/health) | Check node health | Determine if the node is running and ready to accept SQL connections.
+[`/nodes`](../api/cluster/v2.html#operation/listNodes) | List nodes | Get information on all nodes in the cluster including node IDs, software versions, and hardware.
+[`/nodes/{node_id}/ranges`](../api/cluster/v2.html#operation/listNodeRanges) | List node ranges | For a specified node, obtain details on the ranges that it hosts. 
+[`/ranges/hot`](../api/cluster/v2.html#operation/listHotRanges) | List hot ranges | Get information on ranges receiving a high number of reads or writes.
+[`/ranges/{range_id}`](../api/cluster/v2.html#operation/listRange) | Get range details | Get detailed technical information on a range. Typically of use only to Cockroach Labs engineers.
+[`/sessions`](../api/cluster/v2.html#operation/listSessions) | List sessions | Get details of all current users' SQL sessions or sessions by a specific user.
+[`/login`](../api/cluster/v2.html#operation/login) | Log in | Authenticate as an admin on the cluster to retrieve a session token to use with further API calls.
+[`/logout`](../api/cluster/v2.html#operation/logout) | Log out | Invalidate the session token.
+
+## Requirements
+
+To connect with the API on a secure cluster, you will need:
+
+* A [client TLS certificate generated for the SQL role](cockroach-cert.html#create-the-certificate-and-key-pair-for-a-client), using the same CA cert used by the cluster. 
```

Comment:
I don't think that's accurate - the TLS cert is only needed to create the session token via the CLI Command, but then it's also not the only way, the HTTP client can also go to the `/login` endpoint and use that to generate a session token.

Once they have a session token, that's the only thing they need to access the API.

Reply:
@knz Thanks for catching this. It's my mistake; in testing, I see I'm required to specify the _CA cert_ for each call; the /login call and every endpoint after that.

Will it be accurate to say that a requirement is that 'the API client must have a copy of the CA cert used to configure the cluster and other clients'?

Reply:
Why is the CA certificate necessary? what breaks if you don't have it?

Reply:
Below is what I get in testing - see the API call without, and then with, specifying the CA cert.

From a Slack chat (pasted further below) I thought @itsbilal indicated this was required for all secure calls, though it's possible that there was a misunderstanding. I do know that it was always included in calls during [acceptance testing](https://docs.google.com/document/d/1bj7_DNSLZTvH0SLpWA4DT8SukD0y9DZb6VP3PTdjKNk/edit).

``` bash
ml@mbp ~ % cd crdb/clusters/apitest

ml@mbp apitest % curl -d "username=mike&password=test" -H 'Content-Type: application/x-www-form-urlencoded' --cacert certs/ca.crt https://localhost:8080/api/v2/login/
{"session":"CIGAus6L2ImRCRIQwqYOrikuxijYkyDaARW/5Q=="}%

ml@mbp apitest % TOKEN="CIGAus6L2ImRCRIQwqYOrikuxijYkyDaARW/5Q=="

ml@mbp apitest % curl -H "X-Cockroach-API-Session: $TOKEN" https://localhost:8080/api/v2/sessions/
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.

ml@mbp apitest % curl -H "X-Cockroach-API-Session: $TOKEN" --cacert certs/ca.crt https://localhost:8080/api/v2/sessions/
{"sessions":[{"node_id":1,"username":"root","client_address":"127.0.0.1:50124","application_name":"$ cockroach sql","active_queries":null,"start":"2021-05-13T14:27:37.223786Z","last_active_query":"SHOW database","id":"Fn6mkOmOo5AAAAAAAAAAAQ==","max_alloc_bytes":952320,"last_active_query_anon":"SHOW database"}],"errors":[],"next":"fDB8MHw="}%
```

```
Mike:
I am going to try to add --cacert to the example curl requests - as it seems virtually required per our earlier chat. Would you agree with doing that? 

Bilal:
sure - though I wonder if that’ll convey its optionality for insecure clusters
I guess you could say curl --cacert <CA cert for secure clusters> <rest of the args>
or curl [--cacert <CA cert for secure clusters>] <rest of the args>
though that makes it less copy-paste-able
```

I can't find the "earlier" transcript referenced there.

Reply:
All right now I understand what is going on. This is necessary because in our currently documented usage, crdb uses a non-standard CA for the HTTP port, and obviously `curl` doesn't know anything about it; so you need to tell it.

However in production usage, we have more and more customers who decide to issue their own HTTP certificates using a public, well-recognized CA. Alternatively, they use a HTTPS load balancer that presents a separate TLS cert to http clients than the one configured inside crdb. In that case, more often than not, `curl` knows about the CA already and there is no need to explain it.

I think it would be important to author a call out to explain these nuances. The CA cert in your examples here are a mere artifact of our (arguably, contrived) security setup in  examples. IT's important to emphasize that end-users should use the parameters that are suitable for their deployment, and that they should seek to determine first if their setup uses a standard CA or not, and only seek out a particular file if their HTTP endpoint uses a non-standard CA (such as the one we build in other doc examples).

Reply:
Got it. Thanks. (We'll have to revisit our use of such security examples and their framing in the docs, as we've touched on before.)

Would this work, for this doc?

> [you will need...]
> 
> * The CA cert used by the cluster or any intermediary proxy server, either in the client's cert store as a trusted certificate authority or as a file manually specified by the HTTP request (for example, using curl's [cacert](https://curl.se/docs/manpage.html#--cacert)).

Reply:
Got it. Thanks. (We'll have to revisit our use of such security examples and their framing in the docs, as we've touched on before.)

Would this work, for this doc?

> [you will need...]
> 
> * The CA cert used by the cluster or any intermediary proxy server, either in the client's cert store as a trusted certificate authority or as a file manually specified by the HTTP request (for example, using curl's [cacert](https://curl.se/docs/manpage.html#--cacert)).

Reply:
Updated the PR to match the above. Please approve if it's now correct/acceptable.